packages/automerge-repo/src/bundle.ts

```diff
@@ -0,0 +1,348 @@
+  offset += docIdBytes.length
+
+  // heads
+  dataview.setUint32(offset, bundle.heads.length)
+  offset += 4
+  for (const head of bundle.heads) {
+    const headBytes = uint8ArrayFromHexString(head)
+    if (headBytes.length !== 32)
+      throw new Error(`Invalid head length: ${headBytes.length}`)
+    output.set(headBytes, offset)
+    offset += 32
+  }
+
+  // deps
+  dataview.setUint32(offset, bundle.deps.length)
+  offset += 4
+  for (const dep of bundle.deps) {
+    const depBytes = uint8ArrayFromHexString(dep)
+    if (depBytes.length !== 32)
+      throw new Error(`Invalid dep length: ${depBytes.length}`)
+    output.set(depBytes, offset)
+    offset += 32
+  }
+
+  // data len
+  dataview.setUint32(offset, bundle.data.length)
+  offset += 4
+
+  // bundle offset
+  dataview.setUint32(offset, bundleOffset)
+  offset += 4
+
+  return offset
+}
+
+function decodeBundle(data: Uint8Array): Map<DocumentId, DocBundle> {
+  const dataview = new DataView(data.buffer, data.byteOffset, data.byteLength)
+  let offset = 0
+
+  // Check magic number
+  const magic = data.slice(offset, offset + 4)
+  if (!magic.every((byte, i) => byte === MAGIC[i])) {
+    throw new Error("Invalid bundle: incorrect magic number")
+  }
+  offset += 4
+
+  // Check version
+  const version = dataview.getUint8(offset)
+  if (version !== 1) {
+    throw new Error(`Unsupported bundle version: ${version}`)
+  }
+  offset += 1
+
+  // Read number of documents
+  const numDocs = dataview.getUint32(offset)
+  offset += 4
+
+  // Read document headers
+  const docHeaders: Array<{
+    docId: DocumentId
+    heads: Heads
+    deps: Heads
+    dataLen: number
+    bundleOffset: number
+  }> = []
+
+  for (let i = 0; i < numDocs; i++) {
+    // Read doc ID
+    const docIdLen = dataview.getUint32(offset)
+    offset += 4
+    const docIdBytes = data.slice(offset, offset + docIdLen)
+    offset += docIdLen
+    const docId = binaryToDocumentId(docIdBytes as BinaryDocumentId)
+
+    // Read heads
+    const numHeads = dataview.getUint32(offset)
+    offset += 4
+    const heads: Heads = []
+    for (let j = 0; j < numHeads; j++) {
+      const headBytes = data.slice(offset, offset + 32)
+      offset += 32
+      heads.push(uint8ArrayToHexString(headBytes))
+    }
+
+    // Read deps
+    const numDeps = dataview.getUint32(offset)
+    offset += 4
+    const deps: Heads = []
+    for (let j = 0; j < numDeps; j++) {
+      const depBytes = data.slice(offset, offset + 32)
+      offset += 32
+      deps.push(uint8ArrayToHexString(depBytes))
+    }
+
+    // Read data length
+    const dataLen = dataview.getUint32(offset)
+    offset += 4
+
+    // Read bundle offset
+    const bundleOffset = dataview.getUint32(offset)
+    offset += 4
+
+    docHeaders.push({ docId, heads, deps, dataLen, bundleOffset })
+  }
+
+  // Read document data
+  const docs = new Map<DocumentId, DocBundle>()
+  const dataStart = offset
+
+  for (const header of docHeaders) {
+    const docData = data.slice(
```

Comment:
consider using [TypedArray.subarray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/subarray) to avoid memcopy